input/azure/vm.go

```diff
@@ -0,0 +1,123 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package azure
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v5"
+	"github.com/elastic/assetbeat/input/internal"
+	stateless "github.com/elastic/beats/v7/filebeat/input/v2/input-stateless"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"strings"
+)
+
+type AzureVMInstance struct {
+	ID             string
+	Name           string
+	SubscriptionID string
+	Region         string
+	Tags           map[string]*string
+	Metadata       mapstr.M
+}
+
+func collectAzureVMAssets(ctx context.Context, client *armcompute.VirtualMachinesClient, subscriptionId string, regions []string, resourceGroup string, log *logp.Logger, publisher stateless.Publisher) error {
+
+	instances, err := getAllAzureVMInstances(ctx, client, subscriptionId, regions, resourceGroup)
+	if err != nil {
+		return err
+	}
+
+	assetType := "azure.vm.instance"
+	assetKind := "host"
+	log.Debug("Publishing Azure VM instances")
+
+	for _, instance := range instances {
+		internal.Publish(publisher, nil,
+			internal.WithAssetCloudProvider("azure"),
+			internal.WithAssetRegion(instance.Region),
+			internal.WithAssetAccountID(instance.SubscriptionID),
+			internal.WithAssetKindAndID(assetKind, instance.ID),
+			internal.WithAssetType(assetType),
+			internal.WithAssetMetadata(instance.Metadata),
+		)
+	}
+
+	return nil
+}
+
+func getAllAzureVMInstances(ctx context.Context, client *armcompute.VirtualMachinesClient, subscriptionId string, regions []string, resourceGroup string) ([]AzureVMInstance, error) {
+	var vmInstances []AzureVMInstance
+	pager := client.NewListAllPager(&armcompute.VirtualMachinesClientListAllOptions{StatusOnly: to.Ptr("true")})
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to advance page: %v", err)
+		}
+		for _, v := range page.Value {
+			if wantRegion(v, regions) && wantResourceGroup(v, resourceGroup) {
+				var status string
+				if v.Properties != nil && v.Properties.InstanceView != nil && len(v.Properties.InstanceView.Statuses) > 1 {
+					status = *v.Properties.InstanceView.Statuses[1].DisplayStatus
+				}
+				vmInstance := AzureVMInstance{
+					ID:             *v.ID,
```

Comment:
Is this the same with the one collected by add_cloud_metadata processor in https://github.com/elastic/beats/blob/73774d5449275253850a0b57045c96e6941fd479/libbeat/processors/add_cloud_metadata/provider_azure_vm.go#L43 ? After all that is what matters to match the assets together.

Reply:
Thanks for spotting this. It looks like the add_cloud_metadata processor uses the `vmId`, which is different from the `resourceId` (what I initially chose to use in this PR). I personally think the `resourceId` would be a better and easier identifier to handle. However, we need consistency among our products so I changed `asset.id` to match `vmId` (see https://github.com/elastic/assetbeat/pull/314/commits/4b44eefbd382150c30dc90189fe05a0884e702ad)